File: delta/src/features/visualization/telem/retriever.ts
/*
 * Copyright 2023 Synnax Labs, Inc.
 *
 * Use of this software is governed by the Business Source License included in the file
 * licenses/BSL.txt.
 *
 * As of the Change Date specified in that file, in accordance with the Business Source
 * License, use of this software will be governed by the Apache License, Version 2.0,
 * included in the file licenses/APL.txt.
 */

import { Frame, Synnax, TimeRange } from "@synnaxlabs/client";

import { Range } from "@/features/workspace";

export class FrameRetriever {
  private readonly client: Synnax;

  constructor(client: Synnax) {
    this.client = client;
  }

  async get(req: FrameRetrieverRequest): Promise<Frame> {
    const { range, keys } = req;
    const tr = new TimeRange(range.start, range.end);
    return await this.client.data.readFrame(tr, keys);
  }
}

export interface FrameRetrieverRequest {
  range: Range;
  keys: string[];
}

File: core/pkg/service/channel/retriever.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package channel

import (
	"context"

	"github.com/samber/lo"
	dischannel "github.com/synnaxlabs/synnax/pkg/distribution/channel"
)

// RetrieveByKeys returns the minimal distribution representation of the channels matching
// the provided keys, for consumers (the framer codec and writer) that need only the
// distribution-layer channel metadata. Keys that do not resolve to a channel are omitted
// from the result.
func (s *Service) RetrieveByKeys(ctx context.Context, keys ...Key) ([]dischannel.Channel, error) {
	var chs []Channel
	if err := s.newRetrieve().
		Where(MatchKeys(keys...)).
		Entries(&chs).
		Exec(ctx, nil); err != nil {
		return nil, err
	}
	return lo.Map(chs, func(c Channel, _ int) dischannel.Channel { return c.Distribution() }), nil
}

// ContainsKeys reports whether every provided key resolves to an existing channel.
func (s *Service) ContainsKeys(ctx context.Context, keys ...Key) (bool, error) {
	return s.newRetrieve().Where(MatchKeys(keys...)).Exists(ctx, nil)
}
